### ISSUE TITLE
[css-values-4] Definition of a calculation

### ISSUE THREAD

**Issue by cdoublev** [w3c/csswg-drafts#6802]
> A `calc()` function contains a single calculation, which is a sequence of values [...]

> The `min()` or `max()` functions contain one or more comma-separated calculations [...]

> To **parse a calculation**, given a calculation `values`:
>
> [previous steps...]
> 6. Return the result of simplifying a calculation tree from `values`.

> To **simplify a calculation tree** `root`:
>
> [previous steps...]
> 5. If `root` is a Min or Max node

If `calc()`, `min()`, and other functions are not calculations that contains calculation values, instead of being something that contains a calculation, I do not see how the calculation tree can be identified as a Min or Max node in step 5 of the procedure to simplify a calculation tree.

**Comment by cdoublev** [comment]
Also:

> Math functions are turned into calculation trees depending on the function:
>
> - `calc()`: The internal representation of a `calc()` function is the result of parsing a calculation from its argument.

`calc()` is turned into a calculation tree, but the procedure to serialize a calculation tree does not include a step to wrap it in `calc()`. Obviously, `calc()` is turned into calculation tree but it should stay a math function. But I do not see how it can still be a math function because it is turned into a calculation tree by parsing a calculation from its argument, ie. the calculation values.

**Comment by tabatkins** [comment]
> calc() is turned into a calculation tree, but the procedure to serialize a calculation tree does not include a step to wrap it in calc(). 

Right, it shouldn't, that's handled by the "serialize a math function" algo, step 3.

calc() is not a calculation, it's a math function which *contains* a "calculation" as its sole argument and whose *"internal representation"* is a calculation tree (see the first paragraph of [section 11.10](https://drafts.csswg.org/css-values/#calc-internal). 

When you're serializing CSS values, the CSS value is the math function; it's just that part of serializing that function involves serializing the calculation tree it's associated with.

> If calc(), min(), and other functions are not calculations that contain calculation values, instead of being something that contains a calculation, I do not see how the calculation tree can be identified as a Min or Max node in step 5 of the procedure to simplify a calculation tree.

See [section 11.10](https://drafts.csswg.org/css-values/#calc-internal) again - any math function *other than* a calc() has an internal representation that's a calculation tree whose root node is an operator node with the same name as the function.

See also step 5 of "parse a calculation", which replaces math functions in the calculation tree with their internal representations (more calculation trees), so you get a homogenous tree of operators and numeric values. That's how something like `calc(1px + min(10px, 1em))` gets turned into the calculation tree `(Sum 1px (Min 10px 1em))`.

**Comment by cdoublev** [comment]
My misunderstanding came from the terms/verbs *internal representation* and *representing*, and also from how I implemented matching a component value against a CSS grammar.

In this implementation, component values ​​are matched in order from the deepest to shallowest type (`calc-value` then `calc-product` then `calc-sum` then `calc()`), and "hook" functions are run with the component value(s) matching this type, in order to run specific rules for parsing the corresponding type. I created hook functions to implement parsing a calculation: collecting `Product` in the hook for `calc-product`, collecting `Sum` for `calc-sum`, but then I struggled with simplification, which had to run in the hook for parsing math function types, then the above terms in the related procedures became blurry to me.

Anyway, I found an implementation that seems to match the specification, and all tests for math functions pass, including cases like `calc(1s + 1px)`, `calc(2px / 1px)`, `min(2px, 1%)`, etc. I didn't want to read your answers before that, but thank you of course.
